src/charm.py

```diff
@@ -289,7 +296,9 @@ def _on_certificate_expiring(self, event: CertificateExpiringEvent) -> None:
             return
 
         new_csr = generate_csr(
-            private_key=private_key.encode(),
+            private_key=self.model.get_secret(label=PRIVATE_KEY_SECRET_LABEL)
+            .get_content()["private-key"]
```

Comment:
Could/should this code block become a property in the tls-certs library?

For backwards-compatibility you may want a "secret store" wrapper class that chooses a backend (stored or secrets).

Reply:
I've created the secret store, let me know if that works.